Teleport instance lifecycle notes (#8020)
gravitational/teleport: docs/pages/contributing/bootstrapping.mdx

Teleport instance, the `common` module handles setting process defaults,
loading config files and merging this data with the supplied a command-line
to generate a final a proocess config (see `./lib/config`), before handing
control off to the `./lib/service` package.

### 2. `service.Run()`

The `service` wraps the process in a `TeleportProcess` instance. The
`TeleportProcess` acts as a centralised hub for communication between internal
services. It also handles external events like POSIX signals, injecting them
into the internal event bus.

For the most part, this layer consists of configuring a `TeleportProcess`,
starting it and then waiting for it to exit.

Setting up the  the `TeleportProcess` entails things like:

- configuring the main supervisor, the root service that
  starts all other services and acts as an event pump events between
  services. (see `./lib/service.NewSupervisor()`)
- configuring (but not yet _starting_) application services

Once all of the services are configured, they are kicked off by a call to
`TeleportProcess.Start()`.

### 3. `TeleportProcess.Start()`

`TeleportProcess.Start()`

> - Starts the supervisor, and by extension all of the registered services

Reviewer: ```suggestion
- Starts the supervisor, and by extension all of the registered services.
```

> - Wait for signals

Reviewer: ```suggestion
- Waits for signals.
```

> - Exit, potentially forking replacement depending on signal.

Reviewer: Maybe link to this page for how this works? https://goteleport.com/docs/management/admin/upgrading-the-teleport-binary/

## A concrete example: An SSH Node

> ### A Note: The Reverse Tunnel

Reviewer: ```suggestion
### The reverse tunnel
```

The "normal" mode of operation for a node is to

>  1. Connect to the cluster auth server directly, and

Reviewer: Since each item in this list uses a different sentence structure, I think we should switch the list to use regular paragraphs instead.

>  1. Connect to the cluster auth server directly, and

Reviewer: ```suggestion
 1. Connect to the cluster Auth Service directly, and
```

>  2. Proxies make direct connections to the node as and when they're

Reviewer: ```suggestion
 2. Proxy Service instances make direct connections to the Node as and when they're
```

    needed to provide user sessions.

> In some cases (e.g. the node being on the wrong side of a NAT boundary)

Reviewer: While this might change after we revamp the Teleport glossary, we should capitalize "Node" everywhere we're referring to an SSH Service instance to be consistent with our architecture overview (https://goteleport.com/docs/architecture/overview/#definitions).

We should also use "Proxy Service" instead of "proxy" and "Auth Service" instead of "auth server" so this is more explicit that we're referring to the Teleport component/more consistent with the rest of the docs.

the node will have to use a "reverse tunnel" connection instead. This
means that the node will

 1. Connect to a proxy using SSH, and the proxy will forward requests
    on to the auth server, and
 2. User sessions are multiplexed over this tunnel, rather than being
    independent TCP connections.

### Bootstrapping an SSH node

> During the `TeleportProcess` setup, `initSSH()` will register the node

Reviewer: Is the node registration process service the same thing as the `register.node` critical service?

registration process service via `registerWithAuthServer()`. This creates
and registers the `register.node` critical service that will:

- (re)connect to auth server & creates `auth.Client`
- register a process-exit handler that closes the client gracefully
- Broadcast the `SSHIdentity` event once the connection is established,
  with the new client in the payload

> Next, `initSSH()` calls `WaitForEvent`, blocking the setup until until

Reviewer: ```suggestion
Next, `initSSH()` calls `WaitForEvent`, blocking the setup until the
```
